### PR TITLE
Feature/s ht3 value m lmodel

### PR DESCRIPTION
**Pull request by sebastianbierb**
We removed VectorToTensor and TensorToVector, then in valueMLmodel Moritz added a new member which is a Tensor and represents the current state. We followed your suggestion to represent actions by their indices, and implemented successor states accordingly. 
Comments and other cleanup is lacking and will follow later this week after the heat wave passes.

### CODE REVIEW

**Comment by Ung0d** [comment]
Very good. Training for an instance with 74 edges is now fast. Even 973 edges works well.

File: src/ML/pytorch/LinRegMain/linRegMain.cc [hunk 1]
```diff
@@ -33,6 +33,6 @@ int main() {
 	//Start of linear Regression learning with LearnedPolicy by setting agent's policy to lpol
     //agent.setPolicy(&lpol);
     
-    TrainingSet TS(3, 100, 20, 64, 0.1);
+    TrainingSet TS(10, 100, 20, 64, 0.1);
     TS.train(&agent, lpol);
```
Comment (Ung0d) [inline src/ML/pytorch/LinRegMain/linRegMain.cc]: You still use the `rnd` flag when creating the agent. Better directly start with a `LearningPolicy` and also don't create it outside. Use the agents' enum.

File: src/ML/pytorch/valueMLmodel.cc [hunk 2]
```diff
@@ -39,47 +40,21 @@ void valueMLmodel::learn(RLDataset& dataSet, unsigned int numberOfEpochs, unsign
...
-}
-
-// Transforms a tensor of values for the successor states into a vector
-vector<float> valueMLmodel::tensorToVector(torch::Tensor& tens) {
-    vector<float*> vec;
-    vector<float> vecReturn;
-    for (unsigned int i = 0; i < tens.numel(); i++) {
-        vec.push_back(tens[i].data<float>());
-        vecReturn.push_back(*vec.at(i));
+        tensState[i] = (float) s->selectedSubset.at(i);
```
Comment (Ung0d) [inline src/ML/pytorch/valueMLmodel.cc]: I think this can be improved even further, if the tensState from the previous call is reused or (maybe better) if `torch::from_blob` is used in combination with `s->selectedSubset.data()` which directly redirects memory instead of making a copy of the state.

Comment (Ung0d) [reply src/ML/pytorch/valueMLmodel.cc]: If you try `from_blob`: If it works like I think it works (I'm unsure about the conversion from `bool` to `float`), the state and the Tensor `tensState` **share the same memory.** That will make it fast, however, changes to the Tensor will also affect the state. That means the 1's for all successor-states must be switches back to 0 in the Tensor after the predictions are calculated to maintain the state as it was.

File: src/ML/pytorch/valueMLmodel.cc [hunk 3]
```diff
@@ -39,47 +40,21 @@ void valueMLmodel::learn(RLDataset& dataSet, unsigned int numberOfEpochs, unsign
...
+    vector<float*> prediction(dimstate,0);
+    vector<float> prediction1(dimstate,0);
+    torch::Tensor pred = torch::zeros(dimstate);
+    for (unsigned int i = 0; i < index.size(); i++) {
+        tensState[index[i]] = 1;
+        
+        pred = linearNet->forward(tensState);
+        prediction.at(index.at(i)) = pred.data<float>();
+        prediction1.at(index.at(i)) = *prediction.at(index.at(i));
+        tensState[index[i]] = 0;
```
Comment (Ung0d) [inline src/ML/pytorch/valueMLmodel.cc]: Ah, I see, you have to switch back to 0 anyway.

File: src/ML/pytorch/valueMLmodel.cc [hunk 4]
```diff
@@ -39,47 +40,21 @@ void valueMLmodel::learn(RLDataset& dataSet, unsigned int numberOfEpochs, unsign
...
-    vector<float> vecReturn;
-    for (unsigned int i = 0; i < tens.numel(); i++) {
-        vec.push_back(tens[i].data<float>());
-        vecReturn.push_back(*vec.at(i));
+        tensState[i] = (float) s->selectedSubset.at(i);
+        
     }
-    return vecReturn;
+    vector<unsigned int> index = s->calcSuccessorStates();     
+    vector<float*> prediction(dimstate,0);
```
Comment (Ung0d) [inline src/ML/pytorch/valueMLmodel.cc]: Something like `vector<float*>` is very uncommon. I'm not quiet sure what you want to achieve but you probably do it wrong this way.

File: src/ML/pytorch/valueMLmodel.cc [hunk 5]
```diff
@@ -39,47 +40,21 @@ void valueMLmodel::learn(RLDataset& dataSet, unsigned int numberOfEpochs, unsign
...
+    vector<float> prediction1(dimstate,0);
+    torch::Tensor pred = torch::zeros(dimstate);
+    for (unsigned int i = 0; i < index.size(); i++) {
+        tensState[index[i]] = 1;
+        
+        pred = linearNet->forward(tensState);
+        prediction.at(index.at(i)) = pred.data<float>();
+        prediction1.at(index.at(i)) = *prediction.at(index.at(i));
```
Comment (Ung0d) [inline src/ML/pytorch/valueMLmodel.cc]: Just copy the value from the 1x1 Tensor `pred` directly into `prediction1`.

Comment (Ung0d) [reply src/ML/pytorch/valueMLmodel.cc]: Also: I think the softmax is still missing? If you convert the predictions into a `std::vector<float>`, you won't be able to use torch's prebuild softmax function. You could concatenate the 1x1 Tensors `pred` to a (index.size())x1 Tensor and then just call `torch::nn::functional::softmax`. This would be probably more efficient then implementing the softmax yourself.

File: src/ML/pytorch/LinRegMain/linRegMain.cc (hunk 1 above)
Comment (mauricerad) [reply src/ML/pytorch/LinRegMain/linRegMain.cc]: We have tried but met with some difficulty so we focused on this issue first. 
We might make a seperate task for this so it doesn't hinder our progress.

**Comment by Ung0d** [comment]
Deleted because there was another branch made for this
